prow/cmd/deck/pr_history.go

```diff
@@ -0,0 +1,247 @@
+
+func (a byStarted) Len() int           { return len(a) }
+func (a byStarted) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byStarted) Less(i, j int) bool { return a[i].Started.Before(a[j].Started) }
+
+func githubPRLink(org, repo, prNumber string) string {
+	return fmt.Sprintf(githubPRTemplate, org, repo, prNumber)
+}
+
+func githubCommitLink(org, repo, commitHash string) string {
+	return fmt.Sprintf(githubCommitTemplate, org, repo, commitHash)
+}
+
+func jobHistLink(bktName, jobName string) string {
+	return fmt.Sprintf(jobHistLinkTemplate, bktName, jobName)
+}
+
+// gets the pull commit hash from metadata
+func getPullCommitHash(pull string) (string, error) {
+	match := pullCommitRe.FindStringSubmatch(pull)
+	if len(match) != 2 {
+		return "", fmt.Errorf("unable to parse pull %q (expected %q)", pull, expectedPull)
+	}
+	return match[1], nil
+}
+
+func parsePRHistKey(key, defaultOrg, defaultRepo string) (bktName, org, repo, prNumber string, err error) {
+	match := prRe.FindStringSubmatch(key)
+	if len(match) != 5 {
+		err = fmt.Errorf("unrecognized directory layout: %s (expected format %s)", key, expectedPRKey)
+		return
+	}
+	bktName = match[1]
+	org = match[2]
+	if org == "" {
+		org = defaultOrg
+	} else {
+		org = strings.TrimSuffix(org, "_")
+	}
+	repo = match[3]
+	if repo == "" {
+		repo = defaultRepo
+	} else {
+		repo = strings.TrimSuffix(repo, "/")
+	}
+	prNumber = match[4]
+	return
+}
+
+func getPRHistory(url *url.URL, config *config.Config, gcsClient *storage.Client) (prHistoryTemplate, error) {
+	start := time.Now()
+	tmpl := prHistoryTemplate{}
+
+	key := strings.TrimPrefix(url.Path, "/pr-history/")
+	defaultOrg := config.Plank.DefaultDecorationConfig.GCSConfiguration.DefaultOrg
+	defaultRepo := config.Plank.DefaultDecorationConfig.GCSConfiguration.DefaultRepo
+	bktName, org, repo, prNumber, err := parsePRHistKey(key, defaultOrg, defaultRepo)
+	if err != nil {
+		return tmpl, fmt.Errorf("error parsing url: %v", err)
+	}
+	bkt := gcsClient.Bucket(bktName)
+	tmpl.Name = fmt.Sprintf("%s/%s #%s", org, repo, prNumber)
+	tmpl.Link = githubPRLink(org, repo, prNumber)
+
+	p := strings.TrimPrefix(key, bktName+"/")
+	jobNames, err := listSubDirs(bkt, p)
+	if err != nil {
+		return tmpl, fmt.Errorf("failed to get job names: %v", err)
+	}
+
+	commits := make(map[string]*commitData)
+	bd := make([]map[string][]buildData, len(jobNames))
+	width := make([]map[string]int, len(jobNames))
+
+	// concurrently find which jobs have been run on this PR and their build numbers
+	type jobBuilds struct {
+		index  int
+		builds []string
+	}
+	jbch := make(chan jobBuilds)
+	for i, jobPrefix := range jobNames {
+		job := path.Base(jobPrefix)
+		jd := prJobData{
+			Name: job,
+			Link: jobHistLink(bktName, job),
+		}
+		tmpl.Jobs = append(tmpl.Jobs, jd)
+		go func(i int, jobPrefix string) {
+			builds, err := listSubDirs(bkt, jobPrefix)
+			if err != nil {
+				logrus.Warningf("Error getting builds for job %s: %v", job, err)
```

Comment:
Use `WithError` for errors on `logrus` lines. Call out specific items like the job name using `WithField`. All identifying information should be in a separate field to allow for machine-reading and searching these logs